### PR TITLE
Redoes how alarms are handled, moves their behavior to datums

### PR DESCRIPTION
**Pull request by LemonInTheDark**
## About The Pull Request

Adds in a set of datums to support sending, listening and storing alarms
In contrast to the old system, we now store a list of send alarms on the listener, rather then the area itself.

This makes clearing "our" alarms on destroy not a massive headache.
In addition, we now use a direct ref to the area's cameras list and signals to prevent camera hard deletes. This, combined with the aformentioned ability to clear, virtually eliminates hard deletes sourced from alarms caused by strange scenarios like the alarm source moving its area.

This change also means that alarm code is centralized rather then implemented in 5 different locations, which means that making behavior changes is easier, as is adding new things that use the system in some manner. I also converted alarm types to defines, because they were literally just hardcoded

## Why It's Good For The Game

Fixes a singularity sourced harddelete, makes deleting cameras faster since we don't need to iterate all alarm camera lists, and is all in all, good for le game. I found this issue as a part of this harddel fix packet #59996, but I decided to pr it separately because it's not a tweak and is rather a somewhat extensive refactor.

I may have changed behavior slightly, since the old logic for power and atmos alarms involved a bool on the area, I assume to prevent duplicate sends or something. this kinda invalidated the point of the sources list, and I think it was a bit stupid, so I removed the area level duplication prevention, and instead added something similar to prevent sending signals we've already sent to the alarm handler.

Oh, and another behavior change, if an object sends an alarm, it being qdel'd clears it. This isn't a necessity, though it does let me do some things that are pretty clean, but I do think it helps a bit with gameplay. Thousands of atmos alarms all cried out in unison, and then silence.
 
On that topic, alarm logic is kinda scuffed, I think intentionally. Like, if I'm an alert program, I only see alarms that occur after I was created, etc etc. I've tried to maintain that behavior since I think it has soul, but I think you should know for the sake of greater understanding of the system.

### CODE REVIEW

File: code/datums/alert.dm [hunk 1]
```diff
@@ -0,0 +1,181 @@
+///Represents a single source of alerts, one alert handler will only ever count for one alert per listener
+/datum/alert_handler
+	///A list of alert type -> list of area names we currently have alerts in
+	var/list/sent_alerts = list()
+	///Our source atom
+	var/atom/source_atom
+
+/datum/alert_handler/New(atom/_source_atom)
+	if(istype(_source_atom))
+		source_atom = _source_atom
```
Comment (Mothblocks) [inline code/datums/alert.dm:10]: https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#when-passing-vars-through-new-or-initializes-arguments-use-srcvar

File: code/__DEFINES/dcs/signals.dm [hunk 2]
```diff
@@ -45,6 +45,10 @@
 #define COMSIG_WEATHER_START(event_type) "!weather_start [event_type]"
 #define COMSIG_WEATHER_WINDDOWN(event_type) "!weather_winddown [event_type]"
 #define COMSIG_WEATHER_END(event_type) "!weather_end [event_type]"
+/// An alert of some form was sent (datum/alert_handler/source, alert_type, area/source_area))
```
Comment (Mothblocks) [inline code/__DEFINES/dcs/signals.dm:48]:
```suggestion
/// An alert of some form was sent (datum/alert_handler/source, alert_type, area/source_area)
```

Can we also be clear what "alert" means here? From the PR title and from this documentation, I had no idea you were talking about like...air alarms?

File: code/__DEFINES/alert.dm [hunk 3]
```diff
@@ -0,0 +1,7 @@
+//A set of defines to be used by the alert datums
```
Comment (Mothblocks) [inline code/__DEFINES/alert.dm:1]: Doc comment all these.

File: code/__DEFINES/dcs/signals.dm (hunk 2 above)
Comment (Mothblocks) [reply code/__DEFINES/dcs/signals.dm:48]: That comment goes for all your doc comments here.

File: code/datums/alert.dm [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+	else
+		var/source_type = ""
+		if(istype(_source_atom, /datum))
+			source_type = _source_atom.type
+		stack_trace("a non atom was passed into alert_handler! [_source_atom] [source_type]")
+	return ..()
+
+/datum/alert_handler/Destroy()
+	for(var/alert_type in sent_alerts)
+		for(var/area/area_to_clear in sent_alerts[alert_type])
```
Comment (Mothblocks) [inline code/datums/alert.dm:20]: `as anything`?

File: code/datums/alert.dm [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+			//Yeet all connected alerts
+			clear_alert_from_area(alert_type, area_to_clear)
+	source_atom = null
+	return ..()
+
+///Sends an alert to any interested things, does some checks to prevent unneeded work
+///Important to note is that source_atom is not held as a ref, we're used as a proxy to prevent hard deletes
+///optional_camera should only be used when you have one camera you want to pass along to alert listeners, most of the time you should have no use for it
+/datum/alert_handler/proc/send_alert(alert_type, atom/use_as_source_atom = source_atom, optional_camera)
```
Comment (Mothblocks) [inline code/datums/alert.dm:29]: Aren't default arguments that aren't constants like, bad for some reason?

Can you change this to just `use_as_source_atom = use_as_source_atom || source_atom`?

File: code/datums/alert.dm [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+	existing_alerts += our_area
+
+	our_area.active_alarms[alert_type] += 1
+
+	SEND_GLOBAL_SIGNAL(COMSIG_ALERT_FIRE(alert_type), src, alert_type, our_area, our_z_level, optional_camera)
+
+	return TRUE
+
+///Clears an alert from any interested listeners
+/datum/alert_handler/proc/clear_alert(alert_type, use_as_source_atom = source_atom)
```
Comment (Mothblocks) [inline code/datums/alert.dm:56]: Same

File: code/datums/alert.dm [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+	if(!use_as_source_atom)
+		return
+
+	var/area/our_area = get_area(use_as_source_atom)
```
Comment (Mothblocks) [inline code/datums/alert.dm:60]: Shouldn't need the variable.

File: code/datums/alert.dm [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+	///List of allowed areas. if this is null it's ignored
+	var/list/allowed_areas
+
+	///List of alert type -> list of area name -> list(area, ref to area's cameras, list(sources))
+	var/list/alarms = list()
+	///Should we allow alert changes to go through or not
+	var/accepting_alert_changes = TRUE
+
+///Accepts a list of alert types to pay attention to, a list of valid z levels, and a list of valid areas. areas and zlevels are ignored if null
+/datum/alert_listener/New(_alerts_to_listen_for, _allowed_z_levels, _allowed_areas)
```
Comment (Mothblocks) [inline code/datums/alert.dm:98]: https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#when-passing-vars-through-new-or-initializes-arguments-use-srcvar

File: code/datums/alert.dm [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+	allowed_z_levels = _allowed_z_levels
+	allowed_areas = _allowed_areas
+	for(var/alert_type in _alerts_to_listen_for)
+		RegisterSignal(SSdcs, COMSIG_ALERT_FIRE(alert_type), .proc/add_alert)
+		RegisterSignal(SSdcs, COMSIG_ALERT_CLEAR(alert_type), .proc/clear_alert)
+
+	return ..()
+
+///Adds an alert to our alarms list, you probobly shouldn't be calling this manually
```
Comment (Mothblocks) [inline code/datums/alert.dm:107]:
```suggestion
///Adds an alert to our alarms list, you probably shouldn't be calling this manually
```

Can you also describe when you would want to call this manually, and what you'd want to use instead?

File: code/datums/alert.dm [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+	var/list/alerts_of_our_type = alarms[alert_type]
+
+	if(!alerts_of_our_type)
+		return
+
+	if(!alerts_of_our_type[source_area.name])
+		return
+
+	var/list/alarm = alerts_of_our_type[source_area.name]
+	var/list/sources  = alarm[3]
```
Comment (Mothblocks) [inline code/datums/alert.dm:154]:
```suggestion
	var/list/sources = alarm[3]
```

File: code/game/objects/structures/displaycase.dm [hunk 11]
```diff
@@ -98,6 +102,10 @@
 		return
 	var/area/alarmed = get_area(src)
 	alarmed.burglaralert(src)
+
+	alert_manager.send_alert(ALERT_BURGLAR)
+	addtimer(CALLBACK(alert_manager, /datum/alert_handler/proc/clear_alert, ALERT_BURGLAR), 600)
```
Comment (Mothblocks) [inline code/game/objects/structures/displaycase.dm:107]:
```suggestion
	addtimer(CALLBACK(alert_manager, /datum/alert_handler/proc/clear_alert, ALERT_BURGLAR), 60 SECONDS)
```

File: code/modules/mob/living/silicon/ai/ai.dm [hunk 12]
```diff
@@ -522,77 +528,31 @@
...
-		if (cam?.can_use())
-			queueAlarm("--- [class] alarm detected in [home.name]! (<A HREF=?src=[REF(src)];switchcamera=[REF(cam)]>[cam.c_tag]</A>)", class)
-		else if (our_cams?.len)
-			var/foo = 0
-			var/dat2 = ""
-			for (var/obj/machinery/camera/I in our_cams)
-				dat2 += text("[]<A HREF=?src=[REF(src)];switchcamera=[REF(I)]>[]</A>", (!foo) ? "" : " | ", I.c_tag) //I'm not fixing this shit...
-				foo = 1
-			queueAlarm(text ("--- [] alarm detected in []! ([])", class, home.name, dat2), class)
+/mob/living/silicon/ai/proc/alert_triggered(datum/source, alert_type, area/source_area)
```
Comment (Mothblocks) [inline code/modules/mob/living/silicon/ai/ai.dm:531]: SIGNAL_HANDLER

File: code/modules/mob/living/silicon/ai/ai.dm [hunk 13]
```diff
@@ -522,77 +528,31 @@
...
-			if (origin in srcs)
-				srcs -= origin
-			if (srcs.len == 0)
-				cleared = 1
-				L -= I
-	if (cleared)
-		queueAlarm("--- [class] alarm in [A.name] has been cleared.", class, 0)
-		if (viewalerts) ai_alerts()
-	return !cleared
+/mob/living/silicon/ai/proc/alert_cleared(datum/source, alert_type, area/source_area)
```
Comment (Mothblocks) [inline code/modules/mob/living/silicon/ai/ai.dm:552]: SIGNAL_HANDLER

File: code/modules/mob/living/silicon/ai/ai.dm [hunk 14]
```diff
@@ -286,12 +291,13 @@
...
 	dat += "<A HREF='?src=[REF(src)];mach_close=aialerts'>Close</A><BR><BR>"
-	for (var/cat in alarms)
-		dat += text("<B>[]</B><BR>\n", cat)
-		var/list/L = alarms[cat]
-		if (L.len)
-			for (var/alarm in L)
-				var/list/alm = L[alarm]
+	var/list/alarms = listener.alarms
+	for (var/alarm_type in alarms)
+		dat += text("<B>[]</B><BR>\n", alarm_type)
```
Comment (Mothblocks) [inline code/modules/mob/living/silicon/ai/ai.dm:296]: `text` looks...ancient.

```suggestion
		dat += text("<B>[alarm_type]</B><BR>\n")
```

File: code/modules/mob/living/silicon/ai/ai.dm [hunk 15]
```diff
@@ -522,77 +528,31 @@
...
+		if (cam.can_use())
+			queueAlarm("--- [alert_type] alarm detected in [home_name]! (<A HREF=?src=[REF(src)];switchcamera=[REF(cam)]>[cam.c_tag]</A>)", alert_type)
 		else
-			queueAlarm(text("--- [] alarm detected in []! (No Camera)", class, home.name), class)
+			var/first_run = FALSE
+			var/dat2 = ""
+			for (var/obj/machinery/camera/camera as anything in cameras)
+				dat2 += text("[]<A HREF=?src=[REF(src)];switchcamera=[REF(camera)]>[]</A>", (!first_run) ? "" : " | ", camera.c_tag) //I'm not fixing this shit...
+				first_run = TRUE
+			queueAlarm(text ("--- [] alarm detected in []! ([])", alert_type, home_name, dat2), alert_type)
```
Comment (Mothblocks) [inline code/modules/mob/living/silicon/ai/ai.dm:545]: Remove `text`

File: code/modules/mob/living/silicon/ai/ai.dm [hunk 16]
```diff
@@ -522,77 +528,31 @@
...
+	if (length(cameras))
+		var/obj/machinery/camera/cam = cameras[1]
+		if (cam.can_use())
+			queueAlarm("--- [alert_type] alarm detected in [home_name]! (<A HREF=?src=[REF(src)];switchcamera=[REF(cam)]>[cam.c_tag]</A>)", alert_type)
 		else
-			queueAlarm(text("--- [] alarm detected in []! (No Camera)", class, home.name), class)
+			var/first_run = FALSE
+			var/dat2 = ""
+			for (var/obj/machinery/camera/camera as anything in cameras)
+				dat2 += text("[]<A HREF=?src=[REF(src)];switchcamera=[REF(camera)]>[]</A>", (!first_run) ? "" : " | ", camera.c_tag) //I'm not fixing this shit...
```
Comment (Mothblocks) [inline code/modules/mob/living/silicon/ai/ai.dm:543]: Remove `text`

File: code/modules/mob/living/silicon/robot/robot.dm [hunk 17]
```diff
@@ -248,63 +256,11 @@
 	if(connected_ai)
 		. += "Master AI: [connected_ai.name]"
 
+/mob/living/silicon/robot/proc/alert_triggered(datum/source, alert_type, area/source_area)
```
Comment (Mothblocks) [inline code/modules/mob/living/silicon/robot/robot.dm:259]: SIGNAL_HANDLER

File: code/modules/mob/living/silicon/robot/robot.dm [hunk 18]
```diff
@@ -248,63 +256,11 @@
...
-			var/list/srcs  = alarm[3]
-			if (origin in srcs)
-				srcs -= origin
-			if (srcs.len == 0)
-				cleared = TRUE
-				L -= I
-	if (cleared)
-		queueAlarm("--- [class] alarm in [A.name] has been cleared.", class, 0)
-	return !cleared
+/mob/living/silicon/robot/proc/alert_cleared(datum/source, alert_type, area/source_area)
```
Comment (Mothblocks) [inline code/modules/mob/living/silicon/robot/robot.dm:262]: SIGNAL_HANDLER

File: code/modules/mob/living/simple_animal/friendly/drone/_drone.dm [hunk 19]
```diff
@@ -314,60 +321,11 @@
 		adjustBruteLoss(heavy_emp_damage)
 		to_chat(src, span_userdanger("HeAV% DA%^MMA+G TO I/O CIR!%UUT!"))
 
+/mob/living/simple_animal/drone/proc/alert_triggered(datum/source, alert_type, area/source_area)
```
Comment (Mothblocks) [inline code/modules/mob/living/simple_animal/friendly/drone/_drone.dm:324]: SIGNAL_HANDLER

File: code/modules/mob/living/simple_animal/friendly/drone/_drone.dm [hunk 20]
```diff
@@ -314,60 +321,11 @@
...
-				var/list/alarm = L[I]
-				var/list/srcs  = alarm[2]
-				if (origin in srcs)
-					srcs -= origin
-				if (srcs.len == 0)
-					cleared = 1
-					L -= I
-		if(cleared)
-			to_chat(src, "--- [class] alarm in [A.name] has been cleared.")
+/mob/living/simple_animal/drone/proc/alert_cleared(datum/source, alert_type, area/source_area)
```
Comment (Mothblocks) [inline code/modules/mob/living/simple_animal/friendly/drone/_drone.dm:327]: SIGNAL_HANDLER

File: code/modules/modular_computers/file_system/programs/alarm.dm [hunk 21]
```diff
@@ -30,87 +43,19 @@
...
-					L -= I
-			else
-				L -= I // wipe the instances entirely
-				cleared = 1
-
-
-	update_alarm_display()
-	return !cleared
-
 /datum/computer_file/program/alarm_monitor/proc/update_alarm_display()
```
Comment (Mothblocks) [inline code/modules/modular_computers/file_system/programs/alarm.dm:54]: SIGNAL_HANDLER

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/datums/alert.dm (hunk 1 above)
Comment (LemonInTheDark) [reply code/datums/alert.dm:10]: Oh no I forgot the guideline. jesus

File: code/datums/alert.dm (hunk 5 above)
Comment (LemonInTheDark) [reply code/datums/alert.dm]: sure

File: code/__DEFINES/alert.dm (hunk 3 above)
Comment (LemonInTheDark) [reply code/__DEFINES/alert.dm]: o7

File: code/__DEFINES/dcs/signals.dm (hunk 2 above)
Comment (LemonInTheDark) [reply code/__DEFINES/dcs/signals.dm]: I use alert and alarm interchangeably in this pr, it's really more like alarms honestly. I've renamed my code, though honestly now that I've thought about it I'm not sure how clarifying that is.

In general the system is for stuff that wants to alert a lot of other objects in the world when something about it changes, alongside passing along a list of cameras. This is primarily used by alert consoles and silicons. It has some distinguishing features outside of just having a global list of "alarms", primarily supporting new alarm listeners not knowing about old alarms.

Comment (LemonInTheDark) [reply code/__DEFINES/dcs/signals.dm]: I'll elaborate in the datum autodoc

File: code/datums/alert.dm (hunk 7 above)
Comment (LemonInTheDark) [reply code/datums/alert.dm]: I think it aids readability, but I can remove it if you'd like

File: code/datums/alert.dm (hunk 9 above)
Comment (LemonInTheDark) [reply code/datums/alert.dm]: yeah sure

**Comment by Mothblocks** [comment]
im going to get to this

File: code/datums/alarm.dm [hunk 22]
```diff
@@ -0,0 +1,199 @@
...
+	var/atom/source_atom
+
+/datum/alarm_handler/New(atom/source_atom)
+	if(istype(source_atom))
+		src.source_atom = source_atom
+	else
+		var/source_type = ""
+		if(istype(source_atom, /datum))
+			source_type = source_atom.type
+		stack_trace("a non atom was passed into alarm_handler! [source_atom] [source_type]")
```
Comment (Mothblocks) [inline code/datums/alarm.dm:28]:
```suggestion
		stack_trace("a non atom was passed into alarm_handler! [source_atom] [istype(source_atom, /datum) ? source_atom.type : ""]")
```

Comment (Mothblocks) [reply code/datums/alarm.dm:28]: Or if you don't like how much this fits on one line, make the var declaration one line.

File: code/datums/alarm.dm [hunk 23]
```diff
@@ -0,0 +1,199 @@
...
+	src.allowed_areas = allowed_areas
+	for(var/alarm_type in alarms_to_listen_for)
+		RegisterSignal(SSdcs, COMSIG_ALARM_FIRE(alarm_type), .proc/add_alarm)
+		RegisterSignal(SSdcs, COMSIG_ALARM_CLEAR(alarm_type), .proc/clear_alarm)
+
+	return ..()
+
+///Adds an alarm to our alarms list, you shouldn't be calling this manually
+///It should all be handled by the signal listening we do, unless you want to only send an alarm to one listener
+/datum/alarm_listener/proc/add_alarm(datum/source, datum/alarm_handler/handler, alarm_type, area/source_area, source_z, optional_camera)
```
Comment (Mothblocks) [inline code/datums/alarm.dm:125]: SIGNAL_HANDLER

File: code/datums/alarm.dm [hunk 24]
```diff
@@ -0,0 +1,199 @@
...
+		cameras = list(optional_camera) // This will cause harddels, so we need to clear manually
+		RegisterSignal(optional_camera, COMSIG_PARENT_QDELETING, .proc/clear_camera_ref, override = TRUE) //It's just fine to override, cause we clear all refs in the proc
+
+	//This does mean that only the first alarm of that camera type in the area will send a ping, but jesus what else can ya do
+	alarms_of_our_type[source_area.name] = list(source_area, cameras, list(handler))
+	SEND_SIGNAL(src, COMSIG_ALARM_TRIGGERED, alarm_type, source_area)
+
+///Removes an alarm to our alarms list, you probably shouldn't be calling this manually
+///It should all be handled by the signal listening we do, unless you want to only remove an alarm to one listener
+/datum/alarm_listener/proc/clear_alarm(datum/source, datum/alarm_handler/handler, alarm_type, area/source_area)
```
Comment (Mothblocks) [inline code/datums/alarm.dm:159]: SIGNAL_HANDLER

File: code/datums/alarm.dm [hunk 25]
```diff
@@ -0,0 +1,199 @@
...
+	SIGNAL_HANDLER
+	accepting_alarm_changes = FALSE
+
+///Does what it says on the tin, exists for signal hooking
+/datum/alarm_listener/proc/allow_alarm_changes()
+	SIGNAL_HANDLER
+	accepting_alarm_changes = TRUE
+
+///Used to manually clear camera refs if one is ref'd directly
+/datum/alarm_listener/proc/clear_camera_ref(obj/machinery/camera/source)
```
Comment (Mothblocks) [inline code/datums/alarm.dm:192]: SIGNAL_HANDLER

File: code/modules/atmospherics/machinery/airalarm.dm [hunk 26]
```diff
@@ -731,8 +730,14 @@
 	var/new_area_danger_level = 0
 	for(var/obj/machinery/airalarm/AA in A)
 		if (!(AA.machine_stat & (NOPOWER|BROKEN)) && !AA.shorted)
-			new_area_danger_level = clamp(max(new_area_danger_level, AA.danger_level), 0,1)
-	if(A.atmosalert(new_area_danger_level,src)) //if area was in normal state or if area was in alert state
+			new_area_danger_level = clamp(max(new_area_danger_level, AA.danger_level), 0, 1)
+
+	var/did_anything_happen = FALSE
```
Comment (Mothblocks) [inline code/modules/atmospherics/machinery/airalarm.dm:735]:
```suggestion
	var/did_anything_happen
```

It's always set.

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/datums/alarm.dm (hunk 22 above)
Comment (LemonInTheDark) [reply code/datums/alarm.dm:28]: I can do that

**Pull request merged**